core/cli_parser.cpp

```diff
@@ -0,0 +1,943 @@
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2019 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2019 Godot Engine contributors (cf. AUTHORS.md)    */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "cli_parser.h"
+#include "core/os/os.h"
+#include "core/translation.h"
+#include "core/version.h"
+#include "core/version_hash.gen.h"
+#include "main/tests/test_main.h"
+
+#define COMMAND_SEPARATOR "--"
+
+// Checkers
+
+struct CommandFlag::CommandChecker {
+
+	virtual bool check(const String &p_arg) const = 0;
+
+	String error_msg;
+
+	virtual ~CommandChecker() {}
+};
+
+struct CommandFlag::FunctionChecker : public CommandFlag::CommandChecker {
+
+	bool check(const String &p_arg) const override {
+		return function(p_arg);
+	}
+
+	CommandFlag::check_function function;
+};
+
+struct CommandFlag::ObjectChecker : public CommandFlag::CommandChecker {
+
+	bool check(const String &p_arg) const override {
+		Variant::CallError ce;
+		Variant v_arg = (Variant)p_arg;
+		const Variant *arg[1] = { &v_arg };
+
+		bool res = obj->call(func, arg, 1, ce);
+		if (ce.error != Variant::CallError::CALL_OK)
+			res = false;
+
+		return res;
+	}
+
+	StringName func;
+	Object *obj;
+};
```

Comment:
The class used to check the argument passed after a command is polymorphic because it can be a pointer to function (for easier in-engine development), a pointer to Object + StringName (to define checkers from GDScript). A lambda checker could be added in the future as it is a possibility that references to script functions will be added to Variant.